crates/store/src/state/state_db.rs

```diff
@@ -94,152 +291,342 @@ impl<'a> StateTree<'a> {
     }
 
     fn insert_data(&mut self, data_hash: H256, code: Bytes) {
-        self.db()
-            .insert_raw(COLUMN_DATA, data_hash.as_slice(), &code)
-            .expect("insert data");
+        if let Some(state_tracker) = self.state_tracker.as_ref() {
+            state_tracker
+                .write_data()
+                .lock()
+                .unwrap()
+                .insert(data_hash, code.clone());
+        }
+        self.journal.push(JournalEntry::InsertData {
+            data_hash,
+            prev_exist: self.get_data(&data_hash).is_some(),
+        });
+        self.dirty_data.insert(data_hash, code);
     }
 
     fn get_data(&self, data_hash: &H256) -> Option<Bytes> {
-        self.db()
-            .get(COLUMN_DATA, data_hash.as_slice())
-            .map(|slice| Bytes::from(slice.to_vec()))
+        let data = self
+            .dirty_data
+            .get(data_hash)
+            .cloned()
+            .or_else(|| self.state.get_data(data_hash));
+        if let Some(data) = data.as_ref() {
+            if let Some(state_tracker) = self.state_tracker.as_ref() {
+                state_tracker
+                    .read_data()
+                    .lock()
+                    .unwrap()
+                    .insert(*data_hash, data.clone());
+            }
+        }
+        data
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use gw_common::{h256_ext::H256Ext, smt::SMT, state::State, H256};
+    use gw_traits::CodeStore;
+
+    use crate::{
+        smt::smt_store::SMTStateStore,
+        snapshot::StoreSnapshot,
+        state::{
+            overlay::{mem_state::MemStateTree, mem_store::MemStore},
+            traits::JournalDB,
+            MemStateDB,
+        },
+        Store,
+    };
+
+    fn new_state(store: StoreSnapshot) -> MemStateDB {
+        let smt = SMT::new(H256::zero(), SMTStateStore::new(MemStore::new(store)));
+        let inner = MemStateTree::new(smt, 0);
+        MemStateDB::new(inner)
+    }
+
+    fn cmp_dirty_state(state_a: &MemStateDB, state_b: &MemStateDB) -> bool {
+        // account count
+        if state_a.get_account_count() != state_b.get_account_count() {
+            return false;
+        }
+        // scripts
+        for script_hash in state_a
+            .dirty_scripts
+            .keys()
+            .chain(state_b.dirty_scripts.keys())
+        {
+            if state_a.get_script(script_hash) != state_b.get_script(script_hash) {
+                return false;
+            }
+        }
+        // data
+        for data_hash in state_a.dirty_data.keys().chain(state_b.dirty_data.keys()) {
+            if state_a.get_data(data_hash) != state_b.get_data(data_hash) {
+                return false;
+            }
+        }
+        // kv
+        for key in state_a.dirty_state.keys().chain(state_b.dirty_state.keys()) {
+            if state_a.get_raw(key) != state_b.get_raw(key) {
+                return false;
+            }
+        }
+
+        true
+    }
```

Comment:
This method can be implemented much simpler leveraging the `Eq` trait impl for `HashMap`, e.g.:

```rust
    fn cmp_dirty_state(state_a: &MemStateDB, state_b: &MemStateDB) -> bool {
        (
            &state_a.dirty_account_count,
            &state_a.dirty_data,
            &state_a.dirty_scripts,
            &state_a.dirty_state,
        ) == (
            &state_b.dirty_account_count,
            &state_b.dirty_data,
            &state_b.dirty_scripts,
            &state_b.dirty_state,
        )
    }
```

The name could also be a bit more specific, i.e. `dirty_state_eq`.

Reply:
For example:

state_a:
- A -> 1 (mem)
- B -> 2 (mem)

state_b:

- A -> 1 (mem)
- B -> 2 (db)

If we compare dirty state it will return false, but this function expect the return is true.